### PR TITLE
Ad authoring2

### PR DESCRIPTION
**Pull request by amakhdoom**
Input field for landing page url
Input field for landing page type
Drop down for call to action text
Button that displays all the above information when clicked

### CODE REVIEW

File: src/app/ad-authoring/ad-authoring.component.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<div class="ad-authoring">
+    <input type="text" #landingUrl placeholder="Landing Page URL"/><br>
+    <input type="text" #landingType placeholder="Landing Page Type"/><br>
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.component.html:3]: Let's use matInput on these.

File: src/app/ad-authoring/ad-authoring.component.html [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    <select #callToAction>
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.component.html:4]: And mat-select here

https://material.angular.io/components/select/overview

File: src/app/ad-authoring/ad-authoring.component.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        <option value="Shop Now">Shop Now</option>
+        <option value="Apply Now">Apply Now</option>
+        <option value="Book Now">Book Now</option>
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.component.html:7]: We should populate these options using an *ngFor enumerating all the CTA enum values. We can go ahead and define a CTA enum for all the valid CTA values, and define a mapping of CTA enum to display value to use here.

File: src/app/ad-authoring/ad-authoring.state.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  providedIn: 'root'
+})
+export class AdAuthoringStateContainer {
+
+    private state$: BehaviorSubject<AdAuthoringState>;
+
+    protected constructor () {
+        this.state$ = new BehaviorSubject({
+          adAuthorings: [],
+        });
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.state.ts:19]: We can inline the state assignment above where the state$ property is defined and remove the explicit constructor.

File: src/app/app.component.html [hunk 5]
```diff
@@ -1,3 +1,3 @@
-<div class="toolbar" role="banner">
-  AMP Story Ad Authoring
-</div>
+<app-top-bar></app-top-bar>
+
```
Comment (ryanfarrow) [inline src/app/app.component.html:2]: nit: remove the empty line.

File: src/app/top-bar/top-bar.component.scss [hunk 6]
```diff
@@ -0,0 +1,12 @@
+.toolbar {
+    align-items: center;
+    background-color: #1976d2;
+    color: white;
+    display: flex;
+    font-weight: 600;
+    height: 60px;
+    left: 0;
+    // position: absolute;
```
Comment (ryanfarrow) [inline src/app/top-bar/top-bar.component.scss:9]: Remove commented value.

File: src/app/top-bar/top-bar.component.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  selector: 'app-top-bar',
+  templateUrl: './top-bar.component.html',
+  styleUrls: ['./top-bar.component.scss']
+})
+export class TopBarComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit(): void {
+  }
```
Comment (ryanfarrow) [inline src/app/top-bar/top-bar.component.ts:13]: We can remove empty constructor and ngOnInit.

File: src/app/ad-authoring/ad-authoring.state.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { Injectable } from '@angular/core';
+import {Observable, BehaviorSubject} from 'rxjs';
+import {AdAuthoring} from './ad-authoring';
+
+export interface AdAuthoringState {
+  readonly adAuthorings: ReadonlyArray<AdAuthoring>;
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.state.ts:6]: I'm not sure we need an array here, the user will be working on one Ad at a time.

File: src/app/ad-authoring/ad-authoring.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+export interface AdAuthoring {
```
Comment (ryanfarrow) [inline src/app/ad-authoring/ad-authoring.ts:1]: I don't think having a separate interface is necessary for these fields, they can be declared directly in the state.

File: src/app/ad-authoring/ad-authoring.component.html [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            </mat-panel-title>
+        </mat-expansion-panel-header>
+
+        <mat-form-field>
+            <mat-label>Landing Page URL</mat-label>
+            <input matInput 
+                type="text" 
+                #landingUrl 
+                placeholder="Landing Page URL" 
+                [ngModel] = "landingPageUrl"
```
Comment (sushan04) [inline src/app/ad-authoring/ad-authoring.component.html:16]: Can you keep the naming consistent as landingPageUrl or landingUrl across the code?

File: src/app/ad-authoring/ad-authoring.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    await landingUrlInput.setValue('google.com');
+
+    const value = await landingUrlInput.getValue();
+    expect(value).toBe('google.com');
+  });
+
+  it('changing landing url should call updateLandingUrl function', async () => {
+    const landingUrlInput = await loader.getHarness(MatInputHarness);
+
+    await landingUrlInput.setValue('google.com');
```
Comment (sushan04) [inline src/app/ad-authoring/ad-authoring.component.spec.ts:121]: Is 'google.com' the default value? If so, I'd recommend adding https://google.com as AMP requires it.

File: src/app/ad-authoring/ad-authoring.service.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AdAuthoringService {
+  constructor(
+    private readonly adAuthoringState: AdAuthoringWorkflowStateContainer
+  ) {}
+
+  updateLandingUrl(landingUrl: string) {
```
Comment (sushan04) [inline src/app/ad-authoring/ad-authoring.service.ts:18]: Are we going to add any validation around this to make sure the user input is a valid url?

Comment (amakhdoom) [reply src/app/ad-authoring/ad-authoring.service.ts:18]: Added url validation in preview branch

**Pull request merged**